File: orchestrator/src/checks/engine.ts
import { execSync } from 'node:child_process';
import { existsSync, readFileSync } from 'node:fs';
import { resolve } from 'node:path';
import type { ErrorInfo, CheckResult } from '../plan/state.js';

// ── Check output ──

export interface CheckOutput {
  status: 'passed' | 'failed';
  detail: string | null;
  duration_ms: number;
}

export interface CheckRunResult {
  allPassed: boolean;
  results: CheckResult[];
  error: ErrorInfo | null;
}

// ── Secret masking ──

export function maskSecrets(text: string): string {
  return text
    .replace(/(?:sk-|pk-|token_)[a-zA-Z0-9]{20,}/g, '***')
    .replace(/(?:Bearer|Basic)\s+\S{20,}/g, 'Bearer ***')
    .replace(/(?:password|secret|key|token)=\S+/gi, (m) => m.split('=')[0] + '=***');
}

// ── Retryable judgment ──

export function isRetryable(specType: string, detail: string | null): boolean {
  if (specType === 'cmd_succeeds' && detail?.includes('timed out')) {
    return true;
  }
  if (specType === 'wait_until') {
    return true;
  }
  return false;
}

// ── Individual check executors ──

export async function checkFileExists(
  spec: { path: string },
  cwd: string,
): Promise<CheckOutput> {
  const fullPath = resolve(cwd, spec.path);
  const exists = existsSync(fullPath);
  return {
    status: exists ? 'passed' : 'failed',
    detail: exists ? null : `file not found: ${spec.path}`,
    duration_ms: 0,
  };
}

export async function checkCmdSucceeds(
  spec: { cmd: string; timeout_sec?: number },
  cwd: string,
): Promise<CheckOutput> {
  const timeout = (spec.timeout_sec ?? 120) * 1000;
  const start = Date.now();

  try {
    execSync(spec.cmd, {
      cwd,
      timeout,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    return {
      status: 'passed',
      detail: null,
      duration_ms: Date.now() - start,
    };
  } catch (err: unknown) {
    const duration_ms = Date.now() - start;
    const execErr = err as { killed?: boolean; signal?: string; status?: number; stderr?: string; code?: string };

    // Timeout detection: killed flag, ETIMEDOUT code, or signal
    const isTimeout = execErr.killed || execErr.code === 'ETIMEDOUT' || execErr.signal === 'SIGTERM';
    if (isTimeout) {
      return {
        status: 'failed',
        detail: `command timed out after ${spec.timeout_sec ?? 120}s: ${spec.cmd}`,
        duration_ms,
      };
    }

    const stderr = maskSecrets((execErr.stderr ?? '').toString().slice(0, 500));
    return {
      status: 'failed',
      detail: `exit ${execErr.status ?? '?'}: ${stderr}`.trim(),
      duration_ms,
    };
  }
}

export async function checkFileContains(
  spec: { path: string; pattern: string },
  cwd: string,
): Promise<CheckOutput> {
  const fullPath = resolve(cwd, spec.path);
  if (!existsSync(fullPath)) {
    return {
      status: 'failed',
      detail: `file not found: ${spec.path}`,
      duration_ms: 0,
    };
  }
  const content = readFileSync(fullPath, 'utf-8');
  const found = content.includes(spec.pattern);
  return {
    status: found ? 'passed' : 'failed',
    detail: found ? null : `pattern not found in ${spec.path}: "${spec.pattern}"`,
    duration_ms: 0,
  };
}

export async function checkGitClean(
  spec: { allow_untracked?: boolean },
  cwd: string,
): Promise<CheckOutput> {
  const start = Date.now();
  try {
    const output = execSync('git status --porcelain', {
      cwd,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    const lines = output.trim().split('\n').filter(Boolean);
    const dirty = spec.allow_untracked
      ? lines.filter((l) => !l.startsWith('??'))
      : lines;

    return {
      status: dirty.length === 0 ? 'passed' : 'failed',
      detail: dirty.length === 0 ? null : `${dirty.length} dirty files`,
      duration_ms: Date.now() - start,
    };
  } catch {
    return {
      status: 'failed',
      detail: 'git status failed (not a git repo?)',
      duration_ms: 0,
    };
  }
}

export async function checkEddaEvent(
  spec: { event_type: string; after?: string },
  cwd: string,
  phaseStartedAt?: string,
): Promise<CheckOutput> {
  const start = Date.now();
  const after = spec.after === '$phase_start' ? phaseStartedAt : spec.after;
  try {
    const afterFlag = after ? ` --after "${after}"` : '';
    const cmd = `edda log --json --type ${spec.event_type}${afterFlag} --limit 1`;

    const output = execSync(cmd, { cwd, encoding: 'utf-8', stdio: ['pipe', 'pipe', 'pipe'] });
    const events = JSON.parse(output);
    return {
      status: Array.isArray(events) && events.length > 0 ? 'passed' : 'failed',
      detail: Array.isArray(events) && events.length > 0 ? null : `no "${spec.event_type}" event found`,
      duration_ms: Date.now() - start,
    };
  } catch (err: unknown) {
    return {
      status: 'failed',
      detail: `edda log failed: ${err instanceof Error ? err.message : String(err)}`,
      duration_ms: 0,
    };
  }
}

// ── Check type registry ──

type CheckFn = (spec: Record<string, unknown>, cwd: string, phaseStartedAt?: string) => Promise<CheckOutput>;

// Lazy import to avoid circular dependency
async function checkWaitUntilWrapper(spec: Record<string, unknown>, cwd: string): Promise<CheckOutput> {
  const { checkWaitUntil } = await import('./wait-until.js');
  return checkWaitUntil(spec as unknown as Parameters<typeof checkWaitUntil>[0], cwd);
}

const CHECK_REGISTRY: Record<string, CheckFn> = {
  file_exists: checkFileExists as CheckFn,
  cmd_succeeds: checkCmdSucceeds as CheckFn,
  file_contains: checkFileContains as CheckFn,
  git_clean: checkGitClean as CheckFn,
  edda_event: checkEddaEvent as CheckFn,
  wait_until: checkWaitUntilWrapper as CheckFn,
};

export async function runCheck(
  spec: Record<string, unknown>,
  cwd: string,
  phaseStartedAt?: string,
): Promise<CheckOutput> {
  const type = spec.type as string;
  const fn = CHECK_REGISTRY[type];
  if (!fn) {
    return {
      status: 'failed',
      detail: `unknown check type: ${type}`,
      duration_ms: 0,
    };
  }
  return fn(spec, cwd, phaseStartedAt);
}

// ── Batch runner (short-circuit on first failure) ──

export async function runChecks(
  checks: Record<string, unknown>[],
  cwd: string,
  phaseStartedAt?: string,
): Promise<CheckRunResult> {
  const results: CheckResult[] = [];

  for (let i = 0; i < checks.length; i++) {
    const spec = checks[i]!;
    const type = (spec.type as string) ?? 'unknown';
    const output = await runCheck(spec, cwd, phaseStartedAt);

    results.push({
      type,
      status: output.status,
      detail: output.detail,
    });

    if (output.status === 'failed') {
      // Short-circuit: mark remaining checks as waiting
      for (let j = i + 1; j < checks.length; j++) {
        results.push({
          type: (checks[j]!.type as string) ?? 'unknown',
          status: 'waiting',
          detail: null,
        });
      }
      return {
        allPassed: false,
        results,
        error: {
          type: 'check_failed',
          message: output.detail ?? `check "${type}" failed`,
          retryable: isRetryable(type, output.detail),
          check_index: i,
          timestamp: new Date().toISOString(),
        },
      };
    }
  }

  return { allPassed: true, results, error: null };
}

File: crates/edda-conductor/src/check/engine.rs
use crate::check::CheckOutput;
use crate::plan::schema::CheckSpec;
use crate::state::machine::{CheckResult, CheckStatus, ErrorInfo, ErrorType};
use std::path::PathBuf;

/// Result of running all checks for a phase.
#[derive(Debug)]
pub struct CheckRunResult {
    pub all_passed: bool,
    pub results: Vec<CheckResult>,
    pub error: Option<ErrorInfo>,
}

/// Check engine: runs check specs against the filesystem.
pub struct CheckEngine {
    cwd: PathBuf,
}

impl CheckEngine {
    pub fn new(cwd: PathBuf) -> Self {
        Self { cwd }
    }

    /// Run all checks in order. Short-circuit on first failure.
    pub async fn run_all(
        &self,
        checks: &[CheckSpec],
        phase_started_at: Option<&str>,
    ) -> CheckRunResult {
        let mut results = Vec::new();

        for (i, spec) in checks.iter().enumerate() {
            let output = self.run_one(spec, phase_started_at).await;
            let status = if output.passed {
                CheckStatus::Passed
            } else {
                CheckStatus::Failed
            };

            results.push(CheckResult {
                check_type: spec.type_name().to_string(),
                status,
                detail: output.detail.clone(),
                duration_ms: output.duration.as_millis() as u64,
            });

            if !output.passed {
                // GH-529: a harness-side timeout must not be classified as a
                // retryable check failure — a retry cannot change the outcome.
                let (error_type, retryable) = if output.timed_out {
                    (ErrorType::Timeout, false)
                } else {
                    (ErrorType::CheckFailed, spec.is_retryable())
                };
                // Mark remaining as Waiting
                for check in &checks[(i + 1)..] {
                    results.push(CheckResult {
                        check_type: check.type_name().to_string(),
                        status: CheckStatus::Waiting,
                        detail: None,
                        duration_ms: 0,
                    });
                }
                return CheckRunResult {
                    all_passed: false,
                    results,
                    error: Some(ErrorInfo {
                        error_type,
                        message: output
                            .detail
                            .unwrap_or_else(|| format!("check {} failed", spec.type_name())),
                        retryable,
                        check_index: Some(i),
                        timestamp: now_rfc3339(),
                    }),
                };
            }
        }

        CheckRunResult {
            all_passed: true,
            results,
            error: None,
        }
    }

    /// Run a single check spec.
    async fn run_one(&self, spec: &CheckSpec, phase_started_at: Option<&str>) -> CheckOutput {
        match spec {
            CheckSpec::FileExists { path } => {
                crate::check::file_exists::check_file_exists(path, &self.cwd)
            }
            CheckSpec::CmdSucceeds { cmd, timeout_sec } => {
                crate::check::cmd_succeeds::check_cmd_succeeds(cmd, *timeout_sec, &self.cwd).await
            }
            CheckSpec::FileContains { path, pattern } => {
                crate::check::file_contains::check_file_contains(path, pattern, &self.cwd)
            }
            CheckSpec::GitClean { allow_untracked } => {
                crate::check::git_clean::check_git_clean(*allow_untracked, &self.cwd).await
            }
            CheckSpec::EddaEvent { event_type, after } => {
                let after_val = after.as_deref().map(|a| {
                    if a == "$phase_start" {
                        phase_started_at.unwrap_or("")
                    } else {
                        a
                    }
                });
                crate::check::edda_event::check_edda_event(event_type, after_val, &self.cwd).await
            }
            CheckSpec::WaitUntil {
                check,
                interval_sec,
                timeout_sec,
                backoff,
            } => {
                crate::check::wait_until::check_wait_until(
                    check,
                    *interval_sec,
                    *timeout_sec,
                    *backoff,
                    &self.cwd,
                    phase_started_at,
                )
                .await
            }
        }
    }
}

fn now_rfc3339() -> String {
    time::OffsetDateTime::now_utc()
        .format(&time::format_description::well_known::Rfc3339)
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::plan::schema::CheckSpec;

    #[tokio::test]
    async fn empty_checks_pass() {
        let dir = tempfile::tempdir().unwrap();
        let engine = CheckEngine::new(dir.path().to_path_buf());
        let result = engine.run_all(&[], None).await;
        assert!(result.all_passed);
        assert!(result.results.is_empty());
    }

    #[tokio::test]
    async fn file_exists_check_passes() {
        let dir = tempfile::tempdir().unwrap();
        std::fs::write(dir.path().join("test.txt"), "hello").unwrap();

        let engine = CheckEngine::new(dir.path().to_path_buf());
        let checks = vec![CheckSpec::FileExists {
            path: "test.txt".into(),
        }];
        let result = engine.run_all(&checks, None).await;
        assert!(result.all_passed);
        assert_eq!(result.results.len(), 1);
        assert_eq!(result.results[0].status, CheckStatus::Passed);
    }

    #[tokio::test]
    async fn file_exists_check_fails() {
        let dir = tempfile::tempdir().unwrap();
        let engine = CheckEngine::new(dir.path().to_path_buf());
        let checks = vec![CheckSpec::FileExists {
            path: "nonexistent.txt".into(),
        }];
        let result = engine.run_all(&checks, None).await;
        assert!(!result.all_passed);
        assert!(result.error.is_some());
    }

    #[tokio::test]
    async fn short_circuit_on_failure() {
        let dir = tempfile::tempdir().unwrap();
        let engine = CheckEngine::new(dir.path().to_path_buf());

        let checks = vec![
            CheckSpec::FileExists {
                path: "nonexistent.txt".into(),
            },
            CheckSpec::FileExists {
                path: "also-missing.txt".into(),
            },
        ];
        let result = engine.run_all(&checks, None).await;
        assert!(!result.all_passed);
        assert_eq!(result.results.len(), 2);
        assert_eq!(result.results[0].status, CheckStatus::Failed);
        assert_eq!(result.results[1].status, CheckStatus::Waiting); // short-circuited
    }

    /// GH-529: a timed-out cmd_succeeds must classify as `ErrorType::Timeout`
    /// with `retryable: false`, distinct from a genuine check failure.
    #[tokio::test]
    async fn cmd_timeout_classifies_as_timeout_not_check_failed() {
        let dir = tempfile::tempdir().unwrap();
        let engine = CheckEngine::new(dir.path().to_path_buf());

        // A command that never finishes inside its 1s budget.
        #[cfg(windows)]
        let cmd = "while ($true) { Start-Sleep -Milliseconds 100 }";
        #[cfg(not(windows))]
        let cmd = "sleep 30";
        let checks = vec![CheckSpec::CmdSucceeds {
            cmd: cmd.into(),
            timeout_sec: 1,
        }];

        let result = engine.run_all(&checks, None).await;
        assert!(!result.all_passed);
        let err = result.error.expect("timeout must carry an error");
        assert_eq!(err.error_type, ErrorType::Timeout);
        assert!(
            !err.retryable,
            "retrying a timeout cannot change the outcome"
        );
        assert!(err.message.contains("timed out"), "got: {}", err.message);
        assert_eq!(result.results[0].status, CheckStatus::Failed);
    }

    /// GH-529: a genuine command failure (non-zero exit) must STILL classify
    /// as `ErrorType::CheckFailed` and stay retryable.
    #[tokio::test]
    async fn cmd_failure_stays_retryable_check_failed() {
        let dir = tempfile::tempdir().unwrap();
        let engine = CheckEngine::new(dir.path().to_path_buf());

        #[cfg(windows)]
        let cmd = "exit 1";
        #[cfg(not(windows))]
        let cmd = "false";
        let checks = vec![CheckSpec::CmdSucceeds {
            cmd: cmd.into(),
            timeout_sec: 10,
        }];

        let result = engine.run_all(&checks, None).await;
        assert!(!result.all_passed);
        let err = result.error.expect("failure must carry an error");
        assert_eq!(err.error_type, ErrorType::CheckFailed);
        assert!(err.retryable);
    }
}
